client/scripts/build-tests.js

```diff
@@ -1,49 +1,167 @@
+
+  let entrypoint = "import '../scripts/tests-shim.js';\n"
+
+  for (const file of sourceFiles) {
+    if (EXCLUDE.some(regex => regex.test(file))) continue
+    entrypoint += `import ${JSON.stringify(`../${file}`)};\n`
+  }
+
+  // actually import the test files
+  for (const file of testFiles) {
+    entrypoint += `import ${JSON.stringify(`../${file}`)};\n`
+  }
+
+  return entrypoint
 }
 
 async function build() {
-  const tests = await getTests()
-  console.log(`[tests] Compiling ${Object.keys(tests).length} files...`)
-  await esbuild.build({
-    // add other modules here if needed
-    external: ["jsdom", "global-jsdom"],
-    inject: ["./scripts/tests-shim.js"],
-    outdir: "tests-dist",
-    entryPoints: tests,
-    bundle: true,
-    treeShaking: true,
-    minify: false,
-    format: "cjs",
-    platform: "node",
-    sourcemap: false,
-    outExtension: { ".js": ".cjs" },
-    loader: { ".wasm": "file" },
+  console.log(`[tests] Compiling test megabundle...`)
+
+  const entrypoint = await generateMegaBundleEntrypoint()
+  const dir = path.resolve(process.cwd(), "tests-dist")
+  const file = path.resolve(dir, "test-megabundle.ts")
+
+  await fs.outputFile(file, entrypoint)
+
+  await vite.build({
+    root: dir,
+
+    define: {
+      "import.meta.url": '"file://test-megabundle"'
+    },
+
+    css: {
+      preprocessorOptions: {
+        scss: { sourceMapEmbed: true, sourceMapContents: true, sourceMap: true }
+      }
+    },
+
     plugins: [
-      sveltePlugin({
-        compileOptions: { css: true, cssHash: () => "svelte" },
-        // render typescript using esbuild rather than tsc
-        preprocess: [typescript(), sveltePreprocess({ typescript: false })]
+      workerPlugin(),
+      tomlPlugin(),
+      svelte({
+        compilerOptions: { cssHash: () => "svelte" },
+        preprocess: [
+          sveltePreprocess({
+            sass: { sourceMapEmbed: true, sourceMapContents: true, sourceMap: true }
+          })
+        ]
       })
-    ]
+    ],
+
+    build: {
+      lib: {
+        entry: file,
+        name: "test-megabundle",
+        fileName: "test-megabundle",
+        formats: ["cjs"]
+      },
+
+      sourcemap: true,
+      target: "esnext",
+      minify: false,
+      brotliSize: false,
+      cssCodeSplit: false,
+
+      rollupOptions: {
+        plugins: [
+          istanbul({
+            babelrc: false,
+            babelHelpers: "bundled",
+            extensions: [".js", ".jsx", ".es6", ".es", ".mjs", ".ts", ".tsx", ".svelte"]
+          })
+        ],
+        treeshake: false,
+        output: {
+          interop: "esModule",
+          compact: true,
+          preferConst: true,
+          sourcemapExcludeSources: true
+        },
+        external: [
+          /uvu/,
+          // /vendor\/prism\.js/,
+          // codemirror has CJS exports now, so this is ok
+          /^@codemirror/,
+          "lezer-tree",
+          // breaks things if included
```

Comment:
Will this list of external dependencies need to be updated each time another is added?

Reply:
I suppose? It's not too often - only `threads` actually broke something when imported. The rest of them are external'd because it speeds up the instrumentation step a lot.

Reply:
I see - I'd assumed that they'd been externalised because the test runner thought that they were things that the tests should cover and the coverage was inaccurate as a result. What determines if a dependency needs to be here?

Reply:
No actually, Istanbul/`nyc` is pretty smart and only reports coverage for things that you've narrowed down as needing covered or directly marked.

What determines what goes in here is:
1. it breaks something if you bundle it (node built-ins, `threads`)
2. it's huge and works fine if it's not bundled